builtin/stash--helper.c

```diff
@@ -0,0 +1,75 @@
+#include "../cache.h"
+#include "../cache-tree.h"
+#include "../diff.h"
+#include "../diffcore.h"
+#include "../gettext.h"
+#include "../git-compat-util.h"
+#include "../revision.h"
+
+static const char builtin_stash__helper_usage[] = {
+	"Usage: git stash--helper --non-patch <i_tree>"
+};
+
+static void diff_callback(struct diff_queue_struct *q,
+					 struct diff_options *options, void *data)
+{
+	int i;
+    fprintf(stderr,"diff_callback\n");
+	for (i = 0; i < q->nr; i++)
+    {
+		struct diff_filepair *p;
+		struct stat st;
+
+        fprintf(stderr,"diff_callback: %d\n", i);
+		p = q->queue[i];
+		add_to_index(&the_index, p->two->path, &st, 0);
+	}
+}
+
+int stash_non_patch(const char *i_tree,	const char *prefix)
+{
+	int result;
+	struct rev_info rev;
+	const char *wt_prefix = NULL;
+	unsigned char sha1[20];
+	const char *me = "git stash--helper";
+
+    fprintf(stderr,"stash_non_patch\n");
+	read_index(&the_index);
+    fprintf(stderr,"read the index\n");
+	init_revisions(&rev, NULL);
+    fprintf(stderr,"initialised rev\n");
+	setup_revisions(0, NULL, &rev, NULL);
+    fprintf(stderr,"setup rev\n");
+	rev.diffopt.output_format |= DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = diff_callback;
+	rev.diffopt.format_callback_data=&result;
+	//diff_setup_done(&rev.diffopt);
+    fprintf(stderr,"prepared diffopts\n");
```

Comment:
If you are testing this with the default Git Bash, you unfortunately have to `fflush(stderr);` because mintty does not really synchronise `stderr` properly.

Have you considered giving the ["GNU debugger"](https://github.com/git-for-windows/git/wiki/Debugging-Git) a spin?